pynbody/snapshot/gadgethdf.py

```diff
@@ -155,6 +155,200 @@ class _SubfindHdfMultiFileManager(_GadgetHdfMultiFileManager):
+
+        self._file_ptype_slice = {} # will map from gadget particle type to location in pynbody logical file map
+        self._file_family_slice = {}
+        for fam in self._all_families:
+            family_length = 0
+
+
+            # A simpler and more readable version of the code below would be:
+            #
+            # for hdf_group in self._all_hdf_groups_in_family(fam):
+            #     family_length += hdf_group[self._size_from_hdf5_key].size
+            #
+            # However, occasionally we need to know where in the pynbody file map the gadget particle types lie.
+            # (Specifically this is used when loading subfind data.) So we need to expand that out a bit and also
+            # keep track of the slice for each gadget particle type.
+
+            ptype_slice_start = family_slice_start
+
+            for particle_type in self._family_to_group_map[fam]:
+
+                ptype_slice_len = 0
+                for hdf_group in self._hdf_files.iter_particle_groups_with_name(particle_type):
+                    ptype_slice_len += hdf_group[self._hdf_files._size_from_hdf5_key].size
+                self._file_ptype_slice[particle_type] = slice(ptype_slice_start, ptype_slice_start + ptype_slice_len)
+                family_length += ptype_slice_len
+                ptype_slice_start += ptype_slice_len
+
+            self._file_family_slice[fam] = slice(family_slice_start, family_slice_start + family_length)
+            family_slice_start += family_length
+
+        self._num_file_particle = family_slice_start
+        
+    def __init_partial_map(self, take):
+        """ Initialize the partial map for loading specific particles """
+        if not self.partial_load:
+            return
+
+        if not isinstance(take, np.ndarray):
+            raise TypeError(f"index_array must be a NumPy array. Got {type(take)} instead.")
+        if take.ndim != 1:
+            raise ValueError(f"index_array must be one-dimensional. Got {take.ndim} dimensions instead.")
+        
+        # Sort the array to ensure we can check continuity properly
+        # index_array = np.sort(index_array) # assume all indices are unique and sorted
+        
+        self._partial_ptype_slice = {}
+        
+        for ptype_name, global_slice in  self._file_ptype_slice.items():
+            global_min = global_slice.start
+            global_max = global_slice.stop
+
+            # Find particles in this range
+            mask = (take >= global_min) & (take < global_max)
+            particles_in_range = take[mask]
+
+            if particles_in_range.size > 0:
+                min_id = particles_in_range[0]
+                max_id = particles_in_range[-1]
+
+                # Check if indices form a continuous sequence
+                expected_count = max_id - min_id + 1
+                if particles_in_range.size != expected_count:
+                    raise ValueError(f"Non-continuous indices detected in {ptype_name}. "
```

Comment:
Could you explain more what this is about? I don't see a test for it and it's not totally clear when it's expected to be raised.